File: python/examples/03_status_monitor.py
"""Example 3: Continuously print gripper state changes (100 ms polling).

Usage:
    python 03_status_monitor.py /dev/ttyUSB0
    python 03_status_monitor.py COM3
"""
import signal
import sys
import time

from aidin_gripper import Gripper


_stop = False


def _on_signal(_signum, _frame):
    global _stop
    _stop = True


def main() -> int:
    default_port = "COM3" if sys.platform == "win32" else "/dev/ttyUSB0"
    port = sys.argv[1] if len(sys.argv) > 1 else default_port

    signal.signal(signal.SIGINT, _on_signal)

    try:
        with Gripper() as g:
            g.connect(port)
            print(f"Monitoring {port} (CTRL+C to quit)")

            prev_sig = None
            while not _stop:
                s = g.read_state()
                sig = (s.stage, s.object_state, s.position, s.fault,
                       s.current // 6)            # bucket current to dampen jitter
                if sig != prev_sig:
                    print(s)
                    prev_sig = sig
                time.sleep(0.1)
            print("Stopped.")
    except Exception as e:
        print(f"ERROR: {e}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: cpp/examples/03_status_monitor.cpp
// Example 3: Continuously print gripper state changes.
//
// Connects to the gripper and polls state every 100 ms. Prints a line
// each time something interesting changes (position, current, object
// state, or fault).  CTRL+C to stop.
//
// Usage:
//   ./03_status_monitor /dev/ttyUSB0
//   03_status_monitor.exe COM3
#include "aidin/gripper.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <thread>

namespace {
std::atomic<bool> g_stop{false};
void onSignal(int) { g_stop.store(true); }
}

int main(int argc, char** argv) {
    const std::string port = (argc > 1) ? argv[1]
#if defined(_WIN32)
                                        : "COM3";
#else
                                        : "/dev/ttyUSB0";
#endif

    std::signal(SIGINT, onSignal);

    try {
        aidin::Gripper g;
        g.connect(port);
        std::cout << "Monitoring " << port << " (CTRL+C to quit)\n";

        aidin::GripperState prev{};
        bool first = true;
        while (!g_stop.load()) {
            const auto s = g.readState();
            const bool changed =
                first ||
                s.position != prev.position ||
                s.object   != prev.object   ||
                s.fault    != prev.fault    ||
                (std::abs(int(s.current) - int(prev.current)) > 5);
            if (changed) {
                std::cout << s << "\n";
                prev  = s;
                first = false;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        std::cout << "Stopped.\n";
    } catch (const std::exception& e) {
        std::cerr << "ERROR: " << e.what() << "\n";
        return 1;
    }
    return 0;
}
